[ClangImporter] Import C structs with strong ARC pointer members

AdamCmiel opened the pull request:
## Summary

Import C structs containing `__strong` ARC-qualified pointer fields into Swift behind the experimental feature flag `ImportCStructsWithArcFields`. This is the first half of the ARC struct import work; weak fields will follow in a separate PR.

### What this enables

With `-enable-experimental-feature ImportCStructsWithArcFields`, C structs like:

```c
struct StrongsInAStructArc {
  __strong MYObject *_Nonnull myobj;
};
```

are imported as Swift structs with proper retain/release semantics. Previously these were rejected because of their non-trivial copy/destroy.

### Approach

**Loadable with member-wise value witnesses.** The imported structs use `LoadableClangRecordTypeInfo` in IRGen, which iterates fields and emits `objc_retain`/`objc_release` per field for copy and destroy. This keeps the structs loadable (values live in SSA registers) and lets LLVM optimize away balancing retain/release pairs. The emitted IR is equivalent to what Clang's `__copy_constructor_`/`__destructor_` helpers would generate.

**Type bridging for Swift-bridgeable ObjC types.** Strong fields whose ObjC type has a native Swift bridge (e.g. `NSString` to `String`) get a computed property with the bridged type. The stored property is renamed with a `_` prefix and made private; a computed property with the original name provides transparent bridging via getter/setter. The memberwise init also accepts the bridged Swift type.

**Partial import is rejected.** If any non-static, non-type member of an ARC struct fails to import, the entire struct is rejected rather than importing a subset of fields (which would produce incorrect layout).

### Commits

1. **Import C structs with strong ARC pointer members behind feature flag** -- core ClangImporter changes, `ImportTypeKind::RecordField` for `OCL_Strong`, partial-import guard, unavailable/renamed struct support, SIL/IR tests
2. **Bridge strong ARC fields with Swift-bridgeable ObjC types** -- bridging computed properties, init parameter bridging, `StrongsInAStruct` (existing pre-flag struct) becomes importable with bridged `String!` fields

### Test coverage

- `objc_bridging_arc_fields.swift` -- type-checking, print-module interface, unavailable/renamed/C-function tests
- `objc_arc_fields_silgen.swift` -- SIL for strong struct copy/destroy
- `objc_arc_fields_irgen.swift` -- LLVM IR retain/release patterns
- `objc_arc_fields_bridging_silgen.swift` -- SIL for bridged getter/setter/init

LucianoPAlmeida commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3417,7 +3413,28 @@ namespace {
+              decl->getType()->getNullability(Impl.getClangASTContext()))
+        {
+          if (*nullability == clang::NullabilityKind::NonNull) {
+            Impl.addImportDiagnostic(
```
We seen to be missing a test case that exercise this diagnostic.

LucianoPAlmeida replied:
Also is it possible for an objc member to be `weak` and `non-null`? I thought weak was implicit always nullable. e.g. if we try to declare a property `(weak, nonnull)` we get `property attributes 'nonnull' and 'weak' are mutually exclusive`. So this is more a question because, mostly curious on this =]

LucianoPAlmeida commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3417,7 +3413,28 @@ namespace {
+        }
+
+        auto typeToWrap = importedType.getType();
+        result->setInterfaceType(WeakStorageType::get(typeToWrap->isOptional() ? typeToWrap : typeToWrap->wrapInOptionalType(), Impl.SwiftContext));
```
Nit: clang format

LucianoPAlmeida commented on test/Inputs/clang-importer-sdk/usr/include/objc_structs.h:
```diff
@@ -8,10 +8,19 @@ struct StructOfBlocks {
-  __weak NSString *nsstr;
+struct WeakAndNonnull {
+  // Weak pointers cannnnn be annotated _Nonnull
+  __weak MYObject *_Nonnull myobj;
```
Ah I see, I really though this wasn't allowed

AdamCmiel replied:
On clang totally fine - Swift won't be able to represent an interface for this

LucianoPAlmeida replied:
Yes, I just asked because at property level if we try `@property (weak, nonnull) NSString* prop;` we get `property attributes 'nonnull' and 'weak' are mutually exclusive` compiler error. So that is why I thought this wasn't allowed.

drodriguez commented on test/ClangImporter/objc_bridging.swift:
```diff
@@ -57,15 +59,44 @@ func mutablePointerToObjC(_ path: String) throws -> NSString {
-
+  _ = s.nsstr! as Bool // expected-error {{cannot convert value of type 'NSString' to type 'Bool' in coercion}}
+  
   // FIXME: Blocks should also be Unmanaged.
-  _ = sb.block as Bool // expected-error {{cannot convert value of type '@convention(block) () -> Void' to type 'Bool' in coercion}}
+  _ = sb.block as Bool // expected-error {{cannot convert value of type '() -> Void' to type 'Bool' in coercion}}
```
I never understood if this change is "for the better" or we are introducing a unintentional difference.

It might be worth it to mention on the Source Compatibility section of the pitch. I think it does not have any practical effects, but changing the convention from `block` to `swift` might break some code.

AdamCmiel replied:
Ah yeah I'll revisit this

drodriguez commented on test/ClangImporter/objc_bridging.swift:
```diff
@@ -57,15 +59,44 @@ func mutablePointerToObjC(_ path: String) throws -> NSString {
+// CHECK-IDE-TEST:   weak var myobj: @sil_weak MYObject?
+// CHECK-IDE-TEST: }
+
+// CHECK-IDE-TEST: struct WeakAndNonnull {
+// CHECK-IDE-TEST: }
```
I don't know how easy is to do, but besides the warning, this should not import at all. The layout that Swift has access to would be incorrect if we import it partially.

AdamCmiel replied:
Ah gotcha, shouldn't be too difficult. We exclude the whole struct already so could find a way to bail there.

Commit:
ClangImporter: Import C structs with strong ARC pointer members behind feature flag

Copy and destroy use member-wise value witnesses (the existing
LoadableClangRecordTypeInfo path in IRGen) rather than Clang's
generated __copy_constructor_ / __destructor_ helpers. The Clang
helpers operate on addresses, so using them would force the struct to
be address-only; keeping the type loadable lets values live in SSA
registers and lets LLVM optimize away balancing retain/release pairs.
The emitted IR is semantically identical — both paths lower to the
same objc_retain / objc_release runtime calls per field.

AdamCmiel commented:
@swift-ci Please test

AdamCmiel commented:
@swift-ci Please test

AdamCmiel commented:
On ObjC __weak storage representation:
  
7y ago, @rjmccall noted on the [forums](https://forums.swift.org/t/importing-c-structs-with-non-trivially-copyable-i-e-arc-pointer-fields/21756/10) that ObjC `__weak` and Swift `weak` are distinct mechanisms (`objc_storeWeak`/`objc_loadWeakRetained` vs Swift's native side-table scheme), and ideally we'd add a dedicated `ObjCWeakStorageType` and `ReferenceOwnership` to model the difference.

  This implementation uses Swift's existing `WeakStorageType` with `ReferenceOwnership::Weak`, which lowers to `swift_unknownObjectWeak*` at IRGen. Those functions dispatch at runtime to the correct implementation based on the object; for ObjC objects they always route to `objc_*Weak`. This is the same mechanism Swift already uses for weak properties imported from ObjC classes, so it's correct and consistent, just slightly imprecise (the runtime dispatch is
  unnecessary since we know these are ObjC objects). Adding `ObjCWeakStorageType` would be a cleaner model but is a substantial cross-cutting change (new `ReferenceOwnership` kind, reference-storage metaprogramming, SIL, IRGen) and could be done as a follow-up.

On code reuse with trivial C structs:

@rjmccall also suggested [sharing](https://forums.swift.org/t/importing-c-structs-with-non-trivially-copyable-i-e-arc-pointer-fields/21756/8) as much code as possible with the existing implementation for trivial C structs. This implementation does that, extending the existing `VisitRecordDecl` path in ClangImporter rather than forking it, and reuses `LoadableClangRecordTypeInfo` in IRGen, which gets member-wise value witnesses from `RecordTypeInfoImpl`. No new `TypeInfo` subclass is needed because the struct remains loadable; the existing loadable Clang record path handles non-trivial fields correctly once the fields themselves have the right `TypeInfo` (strong references, weak references, etc.).

AdamCmiel commented:
@swift-ci Please test

AdamCmiel commented:
@swift-ci Please test

rjmccall commented:
Using `LoadableClangRecordTypeInfo` works fine for `__strong` fields, but a struct with a `__weak` field is not loadable (we cannot relocate the weak reference without informing the runtime), so that does not work.

> Those functions dispatch at runtime to the correct implementation based on the object; for ObjC objects they always route to objc_*Weak. This is the same mechanism Swift already uses for weak properties imported from ObjC classes, so it's correct and consistent, just slightly imprecise (the runtime dispatch is
unnecessary since we know these are ObjC objects).

Hmm, this still makes me uncomfortable. It raises a constraint that's currently internal to the ObjC and Swift runtimes (which are co-developed on Apple platforms) into a permanent requirement.

rjmccall commented:
Oh, y'know, there's a much stronger reason that would be wrong. We *don't* know that we have an ObjC object here; we just know that the static type of the object is expressible in Objective-C. The unknown-object runtime is dispatching dynamically based on whether the actual objects use Swift reference counting, which is a different condition. Your code will misbehave if the object stored in the ObjC weak reference uses Swift reference counting.

AdamCmiel commented:
As far as I understand it, the existing WeakStorageType + load_weak/store_weak + swift_unknownObjectWeak* path already does the right thing for ObjC classes, and adding a new entry to ReferenceStorage.def would ripple through the whole compiler: new AST type, new SIL instructions, new mangling, serialization format bump, demangling, TypeLayout entries, runtime function stubs, and dozens of explicit switch statements. 

All of that to arrive at the same objc_initWeak call that swift_unknownObjectWeakInit already reaches after a cheap runtime nativeness check.

But maybe I'm missing the point of the `ObjCWeakStorageType`. If the per-field operations can work through the existing weak infrastructure, and `__weak` ARC in C only exists on Apple platforms where Swift objects are ObjC-compatible, what functional gain is there to be had from the new storage type at the cost of a lot of plumbing? 

Addressing the other point, I added another commit making the owning struct of a weak ARC reference address-only and wiring up the Clang-synthesized functions from @ahatanaka's llvm work for the VWT. That should simplify the VW generation greatly but does it not beg the same question about the nativeness of the weak field's reference counting?

Lastly, I believe that the vast majority of our codebase's structs we'd like to import do not use weak references so I'm more keen on splitting the implementation into two PRs, one for strong references and type bridging and a follow-on where we can get into the weeds on weaks if needed.

rjmccall commented:
I think splitting the patch into strong and weak support would make perfect sense.

> All of that to arrive at the same objc_initWeak call that swift_unknownObjectWeakInit already reaches after a cheap runtime nativeness check.

I think you may have missed my comment that that runtime nativeness check is not guaranteed to fail for objects stored in ObjC `__weak` references.

AdamCmiel commented:
@swift-ci Please test

AdamCmiel commented:
Tests are failng on infrastructure errors:

```
20:46:55  ERROR: Issue with creating launcher for agent i-094db0356894c7d26. The agent has not been fully initialized yet
22:01:55  EC2 instance for node linux-x86_64 i-094db0356894c7d26 was terminated
22:02:08  Could not connect to i-094db0356894c7d26 to send interrupt signal to process
```

looks like timeouts for CI jobs:

```
[time] Body did not finish within grace period; terminating with extreme prejudice
```

AdamCmiel commented:
@swift-ci Please test

AdamCmiel commented:
@swift-ci Please test

AdamCmiel commented:
@swift-ci Please test

rjmccall commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2417,6 +2417,7 @@ namespace {
       bool isNonTrivialPtrAuth = false;
+      bool hasArcFields = false;
       // FIXME: We should actually support strong ARC references and similar in
       // C structs. That'll require some SIL and IRGen work, though.
```
You're resolving this FIXME.

rjmccall commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2417,6 +2417,7 @@ namespace {
           };
```
I believe this logic needs to recurse into C++ base classes.

rjmccall commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2425,23 +2426,31 @@ namespace {
-              decl->getLocation());
-          return nullptr;
+          // Check if the non-triviality is due to ARC fields and the
+          // experimental feature flag is enabled.
```
I think what you want to do is, if the declaration is non-trivial to primitive copy/destroy, walk the members (and bases) and recursively consider each non-trivial field, comparing the reason it's non-trivial with the appropriate import flags. Also, since we don't want to import structs with `__weak` members at all, you need to recursively walk into field types that are themselves struct types.

Also, I think (Clang's extended) C has a concept of types that are impossible primitive copy/destroy because of union members, and we need to continue to not import those.

Also we should make sure that this (1) works correctly in C++ mode and (2) isn't overriding user-provided non-triviality in C++. (That is, if the type has a user-provided copy constructor and destructor, we don't really care about looking at subobjects anymore.)

rjmccall commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2774,6 +2790,81 @@ namespace {
+      // For ARC structs, detect fields whose ObjC type has a Swift bridge
+      // (e.g., NSString → String). For each such field, make the stored
+      // property private (with a _ prefix) and add a computed property
+      // with the original name and the bridged Swift type.
```
This makes a lot of sense to me, but can we separate it out and just make this PR focus on correctly importing the structs?

Commit:
ClangImporter: Recursively classify non-trivial C record fields

Replace the flat isNonTrivialDueToAddressDiversifiedPtrAuth check with a
recursive classifyNonTrivialRecord that walks fields, nested structs,
and C++ base classes to determine *why* a record is non-trivial. This
lets us reject __weak fields, non-trivial union members, and unknown
causes while still allowing __strong-only and ptrauth-only records
through their respective feature gates.

Also adds test coverage for nested ARC structs, __weak rejection,
nested __weak, union with strong fields, and struct-containing-union.

Commit:
ClangImporter: Refine diagnostics for non-trivial C record import failures

Add cause-specific diagnostics for weak references and strong ARC
references, replacing the generic "not trivial to copy or destroy"
message with actionable guidance.

Commit:
ClangImporter: Test that ARC-only C++ structs remain loadable

Add a SILGen test verifying that C++ structs whose only non-triviality
comes from __strong ARC fields are imported as loadable. Clang treats
these as trivial for call purposes (HasTrivialSpecialMembersForCall),
so canPassInRegisters() returns true and isCxxNonTrivial is not set.

The classifyNonTrivialRecord guard uses hasNonTrivialCopyConstructor
and hasNonTrivialDestructor in C++ interop mode to ensure types with
real C++ non-triviality still get UserProvidedCxx classification.

AdamCmiel commented:
Changes since last review:

  - Replaced the flat lambda with a recursive classifyNonTrivialRecord that walks fields, nested structs, and C++ bases. Returns a NonTrivialCause enum (Trivial, PtrAuthOnly, ARCStrongOnly, HasWeak, HasCUnion, UserProvidedCxx, Other) so we can reject __weak and non-trivial union members with cause-specific diagnostics.
  - Added two new diagnostics: record_non_trivial_weak_ref ("contains a weak reference and cannot be imported") and record_non_trivial_strong_ref ("contains strong references and cannot be imported without enabling experimental feature ImportCStructsWithArcFields").
  - Added test structs covering nested strong fields, nested weak fields, unions with strong fields, and structs containing such unions. All rejection cases tested.
  - Separated the bridging commit into a follow-on PR.
  - For C++ interop: the classifyNonTrivialRecord guard uses hasNonTrivialCopyConstructor/hasNonTrivialDestructor when C++ interop is enabled to catch types with real C++ non-triviality. In practice, the entire classification block is bypassed for CXXRecordDecls because isNonTrivialToPrimitiveCopy (the outer guard) is a C-level primitive copy flag that Clang does not set on C++ types. ARC-only C++ structs are already loadable because Clang marks
  __strong as "trivial for call" (HasTrivialSpecialMembersForCall in DeclCXX.cpp), so canPassInRegisters() returns true and isCxxNonTrivial is not set. Added a SILGen test verifying this.

AdamCmiel commented:
@swift-ci Please test.

AdamCmiel commented:
@swift-ci please test Linux platform.
